### PR TITLE
Stain normalization

### PR DESCRIPTION
**Pull request by behxyz**
### Description
This PR implement an H&E stain normalization for histopathology images, which contains:
- an H&E stain extractor using deconvolution methods from histopathology images, and
- an H&E stain normalizer using the extracted stain vectors. 

Unlike the previous PR (#1998), which was based on `cupy`, its purely based on `numpy` without any CUDA dependency. It expects `numpy.ndarray` as the input image and returns `numpy.ndarray`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/pathology/transforms/stain/array.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+class ExtractHEStains(Transform):
+    """Class to extract a target stain from an image, using stain deconvolution (see Note).
+
+    Args:
+        tli: transmitted light intensity. Defaults to 240.
+        alpha: tolerance in percentile for the pseudo-min (alpha percentile)
+            and pseudo-max (100 - alpha percentile). Defaults to 1.
+        beta: absorbance threshold for transparent pixels. Defaults to 0.15
+        max_cref: reference maximum stain concentrations for Hematoxylin & Eosin (H&E).
+            Defaults to None.
```
Comment (wyli) [inline monai/apps/pathology/transforms/stain/array.py:28]: the defauts are `np.array([1.9705, 1.0308])`

File: monai/apps/pathology/transforms/stain/array.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        Args:
+            img: uint8 RGB image to perform stain deconvolution of
+
+        Return:
+            he: H&E absorbance matrix for the image (first column is H, second column is E, rows are RGB values)
+        """
+        # reshape image and calculate absorbance
+        image = image.reshape((-1, 3))
+        image = image.astype(np.float32) + 1.0
+        absorbance = -np.log(image.clip(max=self.tli) / self.tli)
```
Comment (wyli) [inline monai/apps/pathology/transforms/stain/array.py:66]: do we need some validation of the input of np.log?

**Review by wyli — COMMENTED** [review]
looks good apart from the minor comments. @Nic-Ma could you please help confirm/approve this PR?

File: monai/apps/pathology/transforms/stain/array.py (hunk 2 above)
Comment (behxyz) [reply monai/apps/pathology/transforms/stain/array.py:66]: That's a good point! We should check if the input is intensity (0-255) and not any image or other scales. Thanks @wyli!

File: docs/source/apps.rst [hunk 3]
```diff
@@ -98,3 +98,15 @@ Clara MMARs
...
     :members:
+
+.. automodule:: monai.apps.pathology.transforms.stain.array
+.. autoclass:: ExtractHEStains
+    :members:
+.. autoclass:: NormalizeHEStains
+    :members:
+
+.. automodule:: monai.apps.pathology.transforms.stain.dictionary
+.. autoclass:: ExtractHEStainsD
```
Comment (Nic-Ma) [inline docs/source/apps.rst:109]: Suggest to use `ExtractHEStainsd` to align with other docs, same as `NormalizeHEStainsD`.

File: monai/apps/pathology/transforms/stain/array.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        beta: float = 0.15,
+        max_cref: Optional[np.ndarray] = None,
+    ) -> None:
+        self.tli = tli
+        self.alpha = alpha
+        self.beta = beta
+
+        self.max_cref = max_cref
+        if self.max_cref is None:
+            self.max_cref = np.array([1.9705, 1.0308])
```
Comment (Nic-Ma) [inline monai/apps/pathology/transforms/stain/array.py:52]: As @wyli commented, set default to `1.9705, 1.0308` directly.

Thanks.

File: monai/apps/pathology/transforms/stain/array.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+    def _deconvolution_extract_stain(self, image: np.ndarray) -> np.ndarray:
+        """Perform Stain Deconvolution and return stain matrix for the image.
+
+        Args:
+            img: uint8 RGB image to perform stain deconvolution of
```
Comment (Nic-Ma) [inline monai/apps/pathology/transforms/stain/array.py:58]: Maybe `of` -> `on` or `for`?

File: monai/apps/pathology/transforms/stain/array.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    reference stain is used. Similarly, if no maximum stain concentrations are provided, a
+    reference maximum stain concentrations matrix is used.
+
+    Args:
+        tli: transmitted light intensity. Defaults to 240.
+        alpha: tolerance in percentile for the pseudo-min (alpha percentile) and
+            pseudo-max (100 - alpha percentile). Defaults to 1.
+        beta: absorbance threshold for transparent pixels. Defaults to 0.15.
+        target_he: target stain matrix. Defaults to None.
+        max_cref: reference maximum stain concentrations for Hematoxylin & Eosin (H&E).
```
Comment (Nic-Ma) [inline monai/apps/pathology/transforms/stain/array.py:126]: Default values of `target_he` and `max_cref` are same as the other comment.

File: monai/apps/pathology/transforms/stain/dictionary.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+    """
+
+    def __init__(
+        self,
+        keys: KeysCollection,
+        tli: float = 240,
+        alpha: float = 1,
+        beta: float = 0.15,
+        max_cref: Optional[np.ndarray] = None,
```
Comment (Nic-Ma) [inline monai/apps/pathology/transforms/stain/dictionary.py:51]: Same as the other comment.

File: monai/apps/pathology/transforms/stain/dictionary.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    """
+
+    def __init__(
+        self,
+        keys: KeysCollection,
+        tli: float = 240,
+        alpha: float = 1,
+        beta: float = 0.15,
+        target_he: Optional[np.ndarray] = None,
```
Comment (Nic-Ma) [inline monai/apps/pathology/transforms/stain/dictionary.py:96]: same comment for default value.

**Review by Nic-Ma — APPROVED** [review]
Looks good to me except for some minor comments.
I didn't check whether the implementation matches some algorithm theory.

Thanks.

File: monai/apps/pathology/transforms/stain/array.py (hunk 2 above)
Comment (behxyz) [reply monai/apps/pathology/transforms/stain/array.py:71]: Now it checks the image values to be between 0 and 255.

**Comment by behxyz** [comment]
@wyli and @Nic-Ma 
Thanks for you comments. I addressed all of them. Please let me know if I missed anything.

**Pull request merged**